### ISSUE TITLE
Inactive backup warning!

### ISSUE THREAD

**Issue by dcurrey** [HandyGuySoftware/dupReport#44]
Not sure if this would be a needed feature or not for most.

I have noticed some duplicati users myself included that run some backupsets on a manual basis.  
Most likely the destination server is not always online.

In order be warned about backups that haven't been completed in X amount of days may be useful

Most of the time one would figure this out in the main reports.  If they bother to read them completely

**Comment by HandyGuySoftware** [comment]
>  ...If they bother to read them completely

Some things I can't program around ;-)

Once a backup set (src/dest pair) has been recorded in the database, dR will always look for emails from that backup set. If it doesn't see them when the report is run (because you haven't run a manual backup or something is preventing the backups from running/completing) you will get the "last seen x days ago" message in the report. I'm not sure where else/how to process and produce that warning. What did you have in mind?

**Comment by dcurrey** [comment]
I was thinking of a separate warning email similar to the way warnoncollect works.  But also send independent emails on report.

Just an extra warning that would stand out from the main report.   This is similar to what crashplan sends me if backups haven't been done in 5 days.  

To execute say have command line option -n
~~~
 if globs.opts['nobackupwarn']:
        globs.db.create_no_backup_warn()
~~~
create_no_backup_warn()  would basically be the sql statement to check for inactive backups create the email and pass it to sendemail()


Here is the result
![image](https://user-images.githubusercontent.com/33069071/34056316-80d49a18-e1a0-11e7-9854-0f2164f55166.png)

**Comment by HandyGuySoftware** [comment]
OK, I get it. 

Once upon a time I had the idea of assigning "owners" to each backup set and put their email in the [src-dest] sections of the .rc file. I run the backups on my kids' laptops and occasionally they forget to leave the machine running long enough for the backup to run. The idea was that if a backupset from their machine hasn't been seen for a certain number of days it would send them a reminder email. That's actually where the idea for the separate [src-dest] sections in the .rc file came from, I just repurposed it for the date/time format stuff and never got around to implementing the nag email. This sounds a lot like that.

I was going to start working on the email purge function (issue #42 ) this weekend. Once I get through that I'll take a closer look at this and see what can be done.

Keep 'em coming!

**Comment by dcurrey** [comment]
Nice!

You shouldn't have to much trouble,   I have some beta code that seems to function properly on my test server.   Only dealing with 1 backupset and about 10 emails but everything seems to work correctly.

Let me know if you want to see any of it.

**Comment by HandyGuySoftware** [comment]
Sure. Can you post any code/snippets you have?

**Comment by dcurrey** [comment]
I will try to match line numbers best I can.

Look for DC around stuff I changed
options.py  Line 43.
~~~
    ('main','warnoncollect','false', True),
    ('main','applyutcoffset','false', True),
    ('main','show24hourtime', 'true', True),
    ('main','purgedbemail','false', True),  # DC
    ('main','nobackupwarn','0', True),  # DC
~~~

dupreport.py line 189
~~~
   if (globs.opts['collect'] or not globs.opts['report']):
        # Either we're just collecting or not just reporting
        # Get new messages on server
        newMessages = globs.inServer.checkForMessages()
        if newMessages > 0:
            # DC Reset purgedbemail flag
            sqlStmt="UPDATE emails SET emailFound = 0"
            globs.db.execSqlStmt(sqlStmt)
~~~


Around line 234
~~~
    globs.log.write(1,'Program completed in {:.3f} seconds. Exiting'.format(time.time() - startTime))

    # DC Check for backups that are inactive.
    if int(globs.opts['nobackupwarn']) > 0:
        globs.db.create_no_backup_warn()

    # DC Purge database emails entries.
    if globs.opts['purgedbemail'] == 'true':
        globs.db.purgedbemail()

    # Bye, bye, bye, BYE, BYE!
    globs.closeEverythingAndExit(0)
~~~

dremail.py line 259
~~~        # DC Update sqlStmt for email found flag
        sqlStmt = "INSERT INTO emails(messageId, sourceComp, destComp, emailTimestamp, \
            deletedFiles, deletedFolders, modifiedFiles, examinedFiles, \
            openedFiles, addedFiles, sizeOfModifiedFiles, sizeOfAddedFiles, sizeOfExaminedFiles, \
            sizeOfOpenedFiles, notProcessedFiles, addedFolders, tooLargeFiles, filesWithError, \
            modifiedFolders, modifiedSymlinks, addedSymlinks, deletedSymlinks, partialBackup, \
            dryRun, mainOperation, parsedResult, verboseOutput, verboseErrors, endTimestamp, \
            beginTimestamp, duration, messages, warnings, errors, emailFound) \
            VALUES \
            ('{}', '{}', '{}', {}, {}, {}, {}, {}, {}, {}, {},{},{},{},{},{},{},{},{},{},{}, \
            '{}', '{}', '{}', '{}', '{}', '{}', '{}', '{}', '{}', '{}', \"{}\", \"{}\", \"{}\",1)".format(mParts['messageId'], \
            mParts['sourceComp'], mParts['destComp'], mParts['emailTimestamp'], sParts['deletedFiles'], \
            sParts['deletedFolders'], sParts['modifiedFiles'], sParts['examinedFiles'], sParts['openedFiles'], \
            sParts['addedFiles'], sParts['sizeOfModifiedFiles'], sParts['sizeOfAddedFiles'], sParts['sizeOfExaminedFiles'], sParts['sizeOfOpenedFiles'], \
            sParts['notProcessedFiles'], sParts['addedFolders'], sParts['tooLargeFiles'], sParts['filesWithError'], \
            sParts['modifiedFolders'], sParts['modifiedSymlinks'], sParts['addedSymlinks'], sParts['deletedSymlinks'], \
            sParts['partialBackup'], sParts['dryRun'], sParts['mainOperation'], sParts['parsedResult'], sParts['verboseOutput'], \
            sParts['verboseErrors'], dParts['endTimestamp'], dParts['beginTimestamp'], \
            sParts['duration'], sParts['messages'], sParts['warnings'], sParts['errors'])

        # DC 
        if (globs.opts['collect'] is True):
             globs.db.execSqlStmt("UPDATE backupsets SET lastBackupstamp=\'{}\' WHERE source=\'{}\' AND destination=\'{}\'".format( mParts['emailTimestamp'],  mParts['sourceComp'], mParts['destComp']))

        globs.log.write(3, 'sqlStmt=[{}]'.format(sqlStmt))
        return sqlStmt
~~~
At the end of dremail.py  You might be able to move this to sendemail() I needed to pass the subject
~~~
    # DC Send no backupwarn email warning
    def sendnobackupEmail(self, subj, msgHtml, msgText = None):

        # Build email message
        msg = MIMEMultipart('alternative')
        msg['Subject'] = subj
        msg['From'] = globs.opts['outsender']
        msg['To'] = globs.opts['outreceiver']

        # Record the MIME types of both parts - text/plain and text/html.
        # Attach parts into message container.
        # According to RFC 2046, the last part of a multipart message is best and preferred.
        # So attach text first, then HTML
        if msgText is not None:
             msgPart = MIMEText(msgText, 'plain')
             msg.attach(msgPart)
        if msgHtml is not None:
             msgPart = MIMEText(msgHtml, 'html')
             msg.attach(msgPart)

        # Send the message via local SMTP server.
        # The encode('utf-8') was added to deal with non-english character sets in emails. See Issue #26 for details
        self.server.sendmail(globs.opts['outsender'], globs.opts['outreceiver'], msg.as_string().encode('utf-8'))
~~~

db.py 134
In the "create table email" sqlstmt add to the end
,emailfound int)
 
In the "create table backupsets: sqlstmt add to the end
, lastBackupstamp real, noBackupwarn int)

db.py line 165
~~~
    def searchForMessage(self, msgID):
        globs.log.write(1,'Database.searchForMessage({})'.format(msgID))
        sqlStmt = "SELECT messageId FROM emails WHERE messageId=\'{}\'".format(msgID)
        dbCursor = self.execSqlStmt(sqlStmt)
        idExists = dbCursor.fetchone()
        if idExists:
            # DC
            sqlStmt = "UPDATE emails SET emailFound = 1 WHERE messageId=\'{}\'".format(msgID)
            self.execSqlStmt(sqlStmt)
            globs.log.write(2,'Message [{}] already in email database'.format(msgID))
            return True

			return False
~~~


At the end of db.py add 
~~~
    # DC Create create_no_backup_warn section
    def create_no_backup_warn(self):
    # Check database for inactive backups
        # Get all source / dest pairs with backup warning enabled
        sqlStmt="SELECT source, destination, noBackupwarn FROM backupsets WHERE noBackupwarn > 0"
        dbCursor = self.execSqlStmt(sqlStmt)
        results = dbCursor.fetchall()
        for row in results:
             source = row[0]
             dest = row[1]
             nobackup = row[2]
             # Now lets see who isn't being backed up
             sqlStmt="SELECT lastBackupstamp FROM backupsets WHERE source = '{}' AND destination ='{}' AND lastBackupstamp <= {} - {} * 86400".format(source,dest,drdatetime.currenttimestamp(),nobackup)
             dbCursor = self.execSqlStmt(sqlStmt)
             results2 = dbCursor.fetchall()

             # Compile emsil reports and send
             for row in results2:
                  timesincebackup = int((drdatetime.currenttimestamp() - row[0]) / 86000)
                  msgText=''
                  subj = "BACKUP WARNING {}{}{} not being backed up".format(source,globs.opts['srcdestdelimiter'],dest)
                  msgHtml='<html><head></head><body><table border={} cellpadding="{}">\n'.format(1,5)
                  msgHtml += '<tr><td bgcolor="#FF0000" align="center" colspan = "11"><b>Backup Warning for {}{}{}</b></td></tr>\n'.format(source,globs.opts['srcdestdelimiter'],dest)
                  msgHtml += '<tr><td align="center"> System has not been backed up in the last {} days!<br>'.format(timesincebackup)
                  msgHtml += "If this computer hasn't been used in {} days, no action is required.</td></tr>\n".format(timesincebackup)
                  msgHtml += '<tr><td bgcolor="#E6E6E6" align="center">Last Backup was on {}!</td></tr>\n'.format(drdatetime.currentdatereal(row[0]))
                  msgHtml += '</table></body></html>'
                  msgText += 'Backup Warning for {}{}{}!\n\n'.format(source,globs.opts['srcdestdelimiter'],dest)
                  msgText += 'System has not been backed up in the last {} days!\n'.format(timesincebackup)
                  msgText += 'Last Backup was on {}!\n\n'.format(drdatetime.currentdatereal(row[0]))
                  msgText += "If this computer hasn't been used in {} days, no action is required.\n".format(timesincebackup)
                  globs.outServer.sendnobackupEmail(subj, msgHtml, msgText)

    # DC Delete all emails flagged as not on mail server!

    def purgedbemail(self):
         sqlStmt="DELETE FROM emails WHERE emailFound = 0"
         self.execSqlStmt(sqlStmt)
         self.dbCommit()
~~~

drdemail.py at end of file You might have better way of doing this it ties back to the nobackupwarn()
~~~
def currenttimestamp():
    curtimestamp = int(time.time())
    return curtimestamp

def currentdatereal(curtimestamp):
    curdatereal = time.strftime("%Y/%m/%d %I:%M:%S %p",time.localtime(curtimestamp))
    return curdatereal
~~~

I think that is all I will update if I find something else

**Comment by dcurrey** [comment]
Added code to pull request!
Edit:  Hope I did that right.   If not I am sorry.

I can repost it here if needed.

**Comment by HandyGuySoftware** [comment]
Added this feature today. Took a look through your code but tried a different approach. Rather than adding fields to the DB I just used the data already available as the program was running. Put operating parameters in the .rc file instead of the DB as that makes it easier for the user to configure. Configurable on a whole-program or per-srcdest basis. Check out the readme.md file at the end of the [report] section and the [source -destination] for available options. Let me know what you think.

**Comment by dcurrey** [comment]
Sounds good with everything being in the rc file.

You might want to double check on the collect option.   If I remember correctly I added what became the lastBackupstamp for a reason.  Originally set this up on version I think was 2.01 when you used the lastdate, lasttime combo.  But I think its needed on check only,  otherwise I think it matches the lastTimestamp

Unfortunately I don't have time today to test the current code. :(

**Comment by HandyGuySoftware** [comment]
No worries, whenever you get a chance.

**Comment by HandyGuySoftware** [comment]
Just re-reading your last comment. Should the inactive check be performed if only running in collect mode (-c)? Right now it runs as part of the reporting routine, i.e. not in -c mode. If it needs to run under -c it's a small change in the existing code, but doable. Should still be able to do it without adding fields to the db, but I'll need to look at that. Just checking to clarify.

**Comment by dcurrey** [comment]
Yea, that is the idea to do the addition check during the -c run since you don't see any report data.   Although a second warning email during normal report mode isn't a bad thing.  

When checking it with normal report mode could the line that reads "No new activity. Last activity on 2017-11-17 at 20:00:16 (30 days ago)" be changed to red when the nobackupwarn is hit?  Just something to make it pop out like the errors and warnings do?.

**Comment by HandyGuySoftware** [comment]
OK, getting clearer now. I can just move the checking routine out of report.py and into the main routine in dupReport.py. It'll take a few more DB queries, but still doable (in my head, at least :-) There should also be a way to specify when if you want it run (-c, -t, or always). 

Adding colors are easy. I'll add that to the next update.

**Comment by HandyGuySoftware** [comment]
Added in latest Beta upload. nobackup warning can be configured on a global or a per-backupset basis. Warning email receiver can be set globally or per backup. 

Check out description for nobackupwarn and nbwsubject in [report] section and [source-destination] sections, as well as -w option.

Let me know what you think. (And what doesn't work ;-)

**Comment by dcurrey** [comment]
So far looks good.

The only think I have found that if you run "dupReport.py -i -c"  Don't ask :)  You get the following error.
~~~
Traceback (most recent call last):
  File "./dupReport.py", line 245, in <module>
    sendNoBackupWarnings()
  File "./dupReport.py", line 111, in sendNoBackupWarnings
    diff = drdatetime.daysSince(latestTimeStamp)
  File "/home/dcurrey/dupReport-2.1.0_Beta/drdatetime.py", line 178, in daysSince
    then = datetime.datetime.fromtimestamp(tsIn)
TypeError: a float is required
~~~

I am not sure why you would need to start with this combination so a workaround may not be needed.   It was just an accident I hit it.

Just wanted to give you a heads up in case it pops up under some other weird circumstance.

**Comment by HandyGuySoftware** [comment]
OK, -i -c now works & fix uploaded. 

Just for giggles I tested if "-i -t" (an even weirder combination) works. It does.

**Comment by HandyGuySoftware** [comment]
Fixed in 2.1.0 Beta
